proposals/0000-unicode-for-string-processing.md

```diff
@@ -0,0 +1,1321 @@
+
+<details><summary>Unicode properties</summary>
+
+We can choose to extend a Unicode property to a grapheme cluster in one of several ways:
+
+- *single-scalar*: Only a character that comprises a single Unicode scalar value can match
+- *first-scalar*: If the first Unicode scalar in a character matches, then the character matches
+- *any-scalar*: If any Unicode scalar in a character matches, then the character matches
+- *all-scalars*: A character matches if and only if all its Unicode scalar members match
+
+With a few guidelines, we can make headway on classifying Unicode properties:
+
+- Numeric-related properties, like `Numeric_Value`, should only apply to single-scalar characters, for the reasons described in the "Digit" character class section, above.
+- Any other properties that directly or approximately correspond to regex or POSIX character classes should use the first-scalar rule. This corresponds with the way `Character.isWhitespace` is implemented and generally matches the perceived categorization of characters.
+- Properties that resolve to a unique Unicode scalar, such as `Name`, should only apply to single-scalar characters.
+- Properties that govern the way Unicode scalars combine into characters, such as `Canonical_Combining_Class`, or are otherwise only relevant when examining specific Unicode data, such as `Age`, should only apply to single-scalar characters.
+- Properties that can naturally apply to a sequence of Unicode scalars, such as `Lowercase_Mapping`, should use an all-scalars approach. This corresponds with the way `Character.isLowercased` and other casing properties are implemented.
+
+In many cases, properties with a *single-scalar* treatment won't match any characters at all, and will only be useful when matching with Unicode scalar semantics. For example, `/\p{Emoji_Modifier}/` matches the five Fitzpatrick skin tone modifier Unicode scalar values that affect the appearance of emoji within a grapheme cluster. When matching with grapheme cluster semantics, no match for the pattern will be found. Using Unicode scalar semantics, however, you can search for all characters that include such a modifier:
+
+```
+let regex = /(?u)\y.+?\p{Emoji_Modifier}.+?\y/
+for ch in "👩🏾‍🚀🚀 👨🏻‍🎤🎸 🧑🏻‍💻📲".matches(of: regex) {
+  print(ch)
+}
+// Prints:
+// 👩🏾‍🚀
+// 👨🏻‍🎤
+// 🧑🏻‍💻
+```
+
+The table below shows our best effort at choosing the right manner of extending.
+
+| Property                            | Extension                     | Notes                             |
+|-------------------------------------|-------------------------------|-----------------------------------|
```

Comment:
Double check with the character properties proposal. It might have left some things un or under-specified.